File: c_files/raylib/breakout.c
#include "raylib.h"
#include "raymath.h"
#include <stdbool.h>

#define ROWS 5
#define COLS 10
#define BRICK_WIDTH 79
#define BRICK_HEIGHT 30
#define BRICK_GAP 2

// BRICKS

typedef struct Brick
{
  Rectangle rect;
  Color color;
  bool active;

} Brick;

typedef struct Bricks
{
  Brick list[ROWS * COLS];

} Bricks;

void bricks_init(Bricks *bricks)
{
  for (int i = 0; i < ROWS; i++)
  {
    for (int j = 0; j < COLS; j++)
    {
      int x = j * (BRICK_WIDTH + BRICK_GAP);
      int y = i * (BRICK_HEIGHT + BRICK_GAP);

      Brick brick = {
          .rect = (Rectangle){x, y, BRICK_WIDTH, BRICK_HEIGHT},
          .active = true,
          .color = GRAY,
      };

      bricks->list[i * COLS + j] = brick;
    }
  }
}

void bricks_draw(Bricks *bricks)
{
  for (int i = 0; i < ROWS * COLS; i++)
  {
    if (bricks->list[i].active)
    {
      DrawRectangleRec(bricks->list[i].rect, bricks->list[i].color);
    }
  }
}

// PADDLE

typedef struct Paddle
{
  Rectangle initial_rect;
  Rectangle rect;
  float speed;
  Color color;

} Paddle;

void paddle_init(Paddle *paddle)
{
  float width = 100.0f;
  float height = 10.0f;
  paddle->initial_rect =
      (Rectangle){GetScreenWidth() * 0.5f - width / 2,
                  GetScreenHeight() - height - 10.0f, width, height};
  paddle->rect = paddle->initial_rect;
  paddle->speed = 300.0f;
  paddle->color = BLACK;
}

void paddle_reset(Paddle *paddle) { paddle->rect = paddle->initial_rect; }

void paddle_draw(Paddle *paddle)
{
  DrawRectangleRec(paddle->rect, paddle->color);
}

void paddle_update(Paddle *paddle)
{
  // move paddle
  if (IsKeyDown(KEY_LEFT))
  {
    paddle->rect.x -= paddle->speed * GetFrameTime();
  }
  if (IsKeyDown(KEY_RIGHT))
  {
    paddle->rect.x += paddle->speed * GetFrameTime();
  }

  // paddle bounds
  paddle->rect.x =
      Clamp(paddle->rect.x, 0, GetScreenWidth() - paddle->rect.width);
}

// BALL

typedef struct Ball
{
  Vector2 initial_pos;
  Vector2 pos;
  float radius;
  float speed;
  Vector2 direction;
  Color color;
  bool active;

} Ball;

void ball_init(Ball *ball, Paddle *paddle)
{
  ball->radius = 10.0f;
  ball->initial_pos = (Vector2){paddle->rect.x + paddle->rect.width / 2,
                                paddle->rect.y - ball->radius * 2};
  ball->pos = ball->initial_pos;
  ball->speed = 400.0f;
  ball->direction = (Vector2){1, 1};
  ball->color = RED;
  ball->active = false;
}

void ball_reset(Ball *ball)
{
  ball->pos = ball->initial_pos;
  ball->direction.x *= -1;
  ball->active = false;
}

void ball_draw(Ball *ball)
{
  DrawCircleV((Vector2){ball->pos.x, ball->pos.y}, ball->radius, ball->color);
}

void ball_update(Ball *ball, Paddle *paddle)
{
  // update ball position
  if (!ball->active)
  {
    ball->pos = (Vector2){paddle->rect.x + paddle->rect.width / 2,
                          paddle->rect.y - ball->radius * 2};
  }

  // activate ball
  if (IsKeyPressed(KEY_SPACE))
  {
    ball->active = !ball->active;
  }

  // move ball
  if (ball->active)
  {
    ball->pos.x += ball->direction.x * ball->speed * GetFrameTime();
    ball->pos.y += ball->direction.y * ball->speed * GetFrameTime();
  }

  // ball bounds
  if (ball->pos.x < ball->radius ||
      ball->pos.x > GetScreenWidth() - ball->radius)
  {
    ball->direction.x *= -1;
  }
  if (ball->pos.y < ball->radius)
  {
    ball->direction.y *= -1;
  }

  // normalize direction vector
  if (ball->direction.x != 0 && ball->direction.y != 0)
  {
    ball->direction = Vector2Normalize(ball->direction);
  }
}

// GAME OVER

typedef struct Game
{
  int lives;
  bool game_over;
  bool game_won;

  Bricks bricks;
  Paddle paddle;
  Ball ball;

} Game;

void game_init(Game *game)
{
  game->lives = 5;
  game->game_over = false;
  game->game_won = false;

  bricks_init(&game->bricks);
  paddle_init(&game->paddle);
  ball_init(&game->ball, &game->paddle);
}

void game_reset(Game *game)
{
  paddle_reset(&game->paddle);
  ball_reset(&game->ball);
}

void game_draw(Game *game)
{
  if (!game->game_over && !game->game_won)
  { // draw lives
    DrawText(TextFormat("%d", game->lives), 20, GetScreenHeight() - 50, 40,
             BLACK);

    bricks_draw(&game->bricks);
    paddle_draw(&game->paddle);
    ball_draw(&game->ball);
  }

  if (game->game_over)
  {
    char text1[] = "GAME OVER";
    float font_size1 = 40;
    float text_width1 = MeasureText(text1, font_size1);
    char text2[] = "press ENTER to restart";
    float font_size2 = 30;
    float text_width2 = MeasureText(text2, font_size2);

    DrawText(text1, GetScreenWidth() * 0.5f - text_width1 / 2,
             GetScreenHeight() * 0.5f - font_size1 / 2, font_size1, BLACK);
    DrawText(text2, GetScreenWidth() * 0.5f - text_width2 / 2,
             GetScreenHeight() * 0.5f - font_size2 / 2 + 100, font_size2,
             BLACK);
  }
  if (game->game_won)
  {
    char text1[] = "YOU WIN";
    float font_size1 = 40;
    float text_width1 = MeasureText(text1, font_size1);
    char text2[] = "press ENTER to restart";
    float font_size2 = 30;
    float text_width2 = MeasureText(text2, font_size2);

    DrawText(text1, GetScreenWidth() * 0.5f - text_width1 / 2,
             GetScreenHeight() * 0.5f - font_size1 / 2, font_size1, BLACK);
    DrawText(text2, GetScreenWidth() * 0.5f - text_width2 / 2,
             GetScreenHeight() * 0.5f - font_size2 / 2 + 100, font_size2,
             BLACK);
  }
}

void game_update(Game *game)
{
  if (game->lives <= 0)
  {
    game->game_over = true;
  }

  // game win condition
  bool remaining;
  for (int i = 0; i < ROWS * COLS; i++)
  {
    remaining = false;

    if (game->bricks.list[i].active)
    {
      remaining = true;
      break;
    }
  }

  if (!remaining)
  {
    game->game_won = true;
  }

  if (!game->game_over && !game->game_won)
  {
    paddle_update(&game->paddle);
    ball_update(&game->ball, &game->paddle);

    // ball collision paddle
    if (CheckCollisionCircleRec(game->ball.pos, game->ball.radius,
                                game->paddle.rect))
    {
      game->ball.direction.y *= -1;
    }

    // ball collision bricks
    for (int i = 0; i < ROWS * COLS; i++)
    {
      if (game->bricks.list[i].active)
      {
        if (CheckCollisionCircleRec(game->ball.pos, game->ball.radius,
                                    game->bricks.list[i].rect))
        {
          game->bricks.list[i].active = false;
          game->ball.direction.y *= -1;
        }
      }
    }

    // update lives
    if (game->ball.pos.y > GetScreenHeight() - game->ball.radius)
    {
      game->lives -= 1;
      game_reset(game);
    }
  }
  else
  {
    if (IsKeyPressed(KEY_ENTER))
    {
      game->lives = 5;
      game->game_over = false;
      game->game_won = false;

      bricks_init(&game->bricks);
      paddle_init(&game->paddle);
      ball_init(&game->ball, &game->paddle);
    }
  }
}

// MAIN

int main(void)
{
  const int SCREEN_WIDTH = 800;
  const int SCREEN_HEIGHT = 600;
  const char SCREEN_TITLE[] = "Breakout";
  const Color SCREEN_BACKGROUND = RAYWHITE;

  InitWindow(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE);

  Game game;

  game_init(&game);

  while (!WindowShouldClose())
  {
    BeginDrawing();
    ClearBackground(SCREEN_BACKGROUND);

    game_update(&game);
    game_draw(&game);

    EndDrawing();
  }

  CloseWindow();

  return 0;
}

File: python_files/raylib/breakout.py
import pyray as p
from dataclasses import dataclass
from random import randint

SCREEN_WIDTH: int = 800
SCREEN_HEIGHT: int = 600
SCREEN_TITLE: str = "Breakout"
SCREEN_BACKGROUND: p.Color = p.RAYWHITE

ROWS: int = 5
COLS: int = 10
BRICK_WIDTH: int = 79
BRICK_HEIGHT: int = 30
BRICK_GAP: int = 2


@dataclass
class Brick:
    rect: p.Rectangle
    color: p.Color


class Bricks:
    def __init__(self) -> None:
        self.list: list[Brick] = []

        for i in range(ROWS):
            for j in range(COLS):
                x: int = j * (BRICK_WIDTH + BRICK_GAP)
                y: int = i * (BRICK_HEIGHT + BRICK_GAP)

                brick: Brick = Brick(
                    rect=p.Rectangle(x, y, BRICK_WIDTH, BRICK_HEIGHT),
                    color=p.GRAY,
                )

                self.list.append(brick)

    def draw(self) -> None:
        for brick in self.list:
            p.draw_rectangle_rec(brick.rect, brick.color)


class Paddle:
    def __init__(self) -> None:
        width: float = 100.0
        height: float = 10.0

        self.initial_rect: p.Rectangle = p.Rectangle(
            p.get_screen_width() / 2 - width / 2,
            p.get_screen_height() - height - 10.0,
            width,
            height,
        )
        self.rect: p.Rectangle = self.initial_rect
        self.speed: float = 300.0
        self.color: p.Color = p.BLACK

    def reset(self) -> None:
        self.rect = self.initial_rect

    def draw(self) -> None:
        p.draw_rectangle_rec(self.rect, self.color)

    def update(self) -> None:
        # move paddle
        if p.is_key_down(p.KeyboardKey.KEY_LEFT):
            self.rect.x -= self.speed * p.get_frame_time()
        if p.is_key_down(p.KeyboardKey.KEY_RIGHT):
            self.rect.x += self.speed * p.get_frame_time()

        # paddle bounds
        self.rect.x = max(0.0, min(self.rect.x, p.get_screen_width() - self.rect.width))


class Ball:
    def __init__(self, paddle: Paddle) -> None:
        self.paddle: Paddle = paddle

        self.radius: float = 10.0
        self.initial_pos: p.Vector2 = p.Vector2(
            self.paddle.rect.x + self.paddle.rect.width / 2,
            self.paddle.rect.y - self.radius - 10.0,
        )
        self.pos: p.Vector2 = self.initial_pos
        self.speed: float = 400.0
        random_x: int = -1 if randint(0, 1) == 0 else 1
        self.direction: p.Vector2 = p.Vector2(random_x, 1)
        self.color: p.Color = p.RED
        self.is_active: bool = False

    def reset(self) -> None:
        self.pos = self.initial_pos
        random_x: int = -1 if randint(0, 1) == 0 else 1
        self.direction: p.Vector2 = p.Vector2(random_x, 1)
        self.is_active: bool = False

    def draw(self) -> None:
        p.draw_circle_v(self.pos, self.radius, self.color)

    def update(self) -> None:
        # update ball position if ball inactive
        if not self.is_active:
            self.pos: p.Vector2 = p.Vector2(
                self.paddle.rect.x + self.paddle.rect.width / 2,
                self.paddle.rect.y - self.radius - 10.0,
            )

        if abs(self.pos.x) > 0.01 and abs(self.pos.y) > 0.01:
            self.direction = p.vector2_normalize(self.direction)

        # activate ball
        if p.is_key_pressed(p.KeyboardKey.KEY_SPACE):
            self.is_active = not self.is_active

        # move ball
        if self.is_active:
            self.pos.x += self.direction.x * self.speed * p.get_frame_time()
            self.pos.y += self.direction.y * self.speed * p.get_frame_time()

        # ball bounds
        if self.pos.x < self.radius or self.pos.x > p.get_screen_width() - self.radius:
            self.direction.x *= -1
        if self.pos.y < self.radius:
            self.direction.y *= -1


class Game:
    def __init__(self) -> None:
        self.lives: int = 5
        self.game_over: bool = False
        self.game_won: bool = False

        self.bricks: Bricks = Bricks()
        self.paddle: Paddle = Paddle()
        self.ball: Ball = Ball(self.paddle)

    def reset(self) -> None:
        self.paddle.reset()
        self.ball.reset()

    def draw(self) -> None:
        # draw lives
        p.draw_text(str(self.lives), 20, p.get_screen_height() - 50, 40, p.BLACK)

        self.bricks.draw()
        self.paddle.draw()
        self.ball.draw()

    def update(self) -> None:
        if not self.game_over and not self.game_won:
            if len(self.bricks.list) <= 0:
                self.game_won = True

            if self.lives <= 0:
                self.game_over = True

            self.paddle.update()
            self.ball.update()

            # ball collision paddle
            if p.check_collision_circle_rec(
                self.ball.pos, self.ball.radius, self.paddle.rect
            ):
                self.ball.direction.y *= -1

            # ball collision bricks
            for brick in self.bricks.list:
                if p.check_collision_circle_rec(
                    self.ball.pos, self.ball.radius, brick.rect
                ):
                    self.ball.direction.y *= -1
                    self.bricks.list.remove(brick)

            # update lives
            if self.ball.pos.y > p.get_screen_height():
                self.lives -= 1
                self.reset()

        else:
            if p.is_key_pressed(p.KeyboardKey.KEY_ENTER):
                self.lives = 5
                self.game_over = False
                self.game_won = False

                self.bricks = Bricks()
                self.paddle = Paddle()
                self.ball = Ball(self.paddle)


def main() -> None:
    p.init_window(SCREEN_WIDTH, SCREEN_HEIGHT, SCREEN_TITLE)

    game: Game = Game()

    while not p.window_should_close():
        p.begin_drawing()
        p.clear_background(SCREEN_BACKGROUND)

        game.draw()
        game.update()

        p.end_drawing()

    p.close_window()


if __name__ == "__main__":
    main()
